File: tdrs-backend/tdpservice/search_indexes/documents/ssp.py
"""Elasticsearch document mappings for SSP submission models."""

from django_elasticsearch_dsl import Document
from django_elasticsearch_dsl.registries import registry
from ..models.ssp import SSP_M1, SSP_M2, SSP_M3, SSP_M6, SSP_M7
from .document_base import DocumentBase


@registry.register_document
class SSP_M1DataSubmissionDocument(DocumentBase, Document):
    """Elastic search model mapping for a parsed SSP M1 data file."""

    class Index:
        """ElasticSearch index generation settings."""

        name = 'ssp_m1_submissions'
        settings = {
            'number_of_shards': 1,
            'number_of_replicas': 0,
        }

    class Django:
        """Django model reference and field mapping."""

        model = SSP_M1
        fields = [
            'RecordType',
            'RPT_MONTH_YEAR',
            'CASE_NUMBER',
            'FIPS_CODE',
            'COUNTY_FIPS_CODE',
            'STRATUM',
            'ZIP_CODE',
            'DISPOSITION',
            'NBR_FAMILY_MEMBERS',
            'FAMILY_TYPE',
            'TANF_ASST_IN_6MONTHS',
            'RECEIVES_SUB_HOUSING',
            'RECEIVES_MED_ASSISTANCE',
            'RECEIVES_FOOD_STAMPS',
            'AMT_FOOD_STAMP_ASSISTANCE',
            'RECEIVES_SUB_CC',
            'AMT_SUB_CC',
            'CHILD_SUPPORT_AMT',
            'FAMILY_CASH_RESOURCES',
            'CASH_AMOUNT',
            'NBR_MONTHS',
            'CC_AMOUNT',
            'CHILDREN_COVERED',
            'CC_NBR_MONTHS',
            'TRANSP_AMOUNT',
            'TRANSP_NBR_MONTHS',
            'TRANSITION_SERVICES_AMOUNT',
            'TRANSITION_NBR_MONTHS',
            'OTHER_AMOUNT',
            'OTHER_NBR_MONTHS',
            'SANC_REDUCTION_AMT',
            'WORK_REQ_SANCTION',
            'FAMILY_SANC_ADULT',
            'SANC_TEEN_PARENT',
            'NON_COOPERATION_CSE',
            'FAILURE_TO_COMPLY',
            'OTHER_SANCTION',
            'RECOUPMENT_PRIOR_OVRPMT',
            'OTHER_TOTAL_REDUCTIONS',
            'FAMILY_CAP',
            'REDUCTIONS_ON_RECEIPTS',
            'OTHER_NON_SANCTION',
            'WAIVER_EVAL_CONTROL_GRPS',
        ]


@registry.register_document
class SSP_M2DataSubmissionDocument(DocumentBase, Document):
    """Elastic search model mapping for a parsed SSP M2 data file."""

    class Index:
        """ElasticSearch index generation settings."""

        name = 'ssp_m2_submissions'
        settings = {
            'number_of_shards': 1,
            'number_of_replicas': 0,
        }

    class Django:
        """Django model reference and field mapping."""

        model = SSP_M2
        fields = [
            'RecordType',
            'RPT_MONTH_YEAR',
            'CASE_NUMBER',
            'FIPS_CODE',
            'FAMILY_AFFILIATION',
            'NONCUSTODIAL_PARENT',
            'DATE_OF_BIRTH',
            'SSN',
            'RACE_HISPANIC',
            'RACE_AMER_INDIAN',
            'RACE_ASIAN',
            'RACE_BLACK',
            'RACE_HAWAIIAN',
            'RACE_WHITE',
            'GENDER',
            'FED_OASDI_PROGRAM',
            'FED_DISABILITY_STATUS',
            'DISABLED_TITLE_XIVAPDT',
            'AID_AGED_BLIND',
            'RECEIVE_SSI',
            'MARITAL_STATUS',
            'RELATIONSHIP_HOH',
            'PARENT_MINOR_CHILD',
            'NEEDS_PREGNANT_WOMAN',
            'EDUCATION_LEVEL',
            'CITIZENSHIP_STATUS',
            'COOPERATION_CHILD_SUPPORT',
            'EMPLOYMENT_STATUS',
            'WORK_ELIGIBLE_INDICATOR',
            'WORK_PART_STATUS',
            'UNSUB_EMPLOYMENT',
            'SUB_PRIVATE_EMPLOYMENT',
            'SUB_PUBLIC_EMPLOYMENT',
            'WORK_EXPERIENCE_HOP',
            'WORK_EXPERIENCE_EA',
            'WORK_EXPERIENCE_HOL',
            'OJT',
            'JOB_SEARCH_HOP',
            'JOB_SEARCH_EA',
            'JOB_SEARCH_HOL',
            'COMM_SERVICES_HOP',
            'COMM_SERVICES_EA',
            'COMM_SERVICES_HOL',
            'VOCATIONAL_ED_TRAINING_HOP',
            'VOCATIONAL_ED_TRAINING_EA',
            'VOCATIONAL_ED_TRAINING_HOL',
            'JOB_SKILLS_TRAINING_HOP',
            'JOB_SKILLS_TRAINING_EA',
            'JOB_SKILLS_TRAINING_HOL',
            'ED_NO_HIGH_SCHOOL_DIPL_HOP',
            'ED_NO_HIGH_SCHOOL_DIPL_EA',
            'ED_NO_HIGH_SCHOOL_DIPL_HOL',
            'SCHOOL_ATTENDENCE_HOP',
            'SCHOOL_ATTENDENCE_EA',
            'SCHOOL_ATTENDENCE_HOL',
            'PROVIDE_CC_HOP',
            'PROVIDE_CC_EA',
            'PROVIDE_CC_HOL',
            'OTHER_WORK_ACTIVITIES',
            'DEEMED_HOURS_FOR_OVERALL',
            'DEEMED_HOURS_FOR_TWO_PARENT',
            'EARNED_INCOME',
            'UNEARNED_INCOME_TAX_CREDIT',
            'UNEARNED_SOCIAL_SECURITY',
            'UNEARNED_SSI',
            'UNEARNED_WORKERS_COMP',
            'OTHER_UNEARNED_INCOME',
        ]


@registry.register_document
class SSP_M3DataSubmissionDocument(DocumentBase, Document):
    """Elastic search model mapping for a parsed SSP M3 data file."""

    class Index:
        """ElasticSearch index generation settings."""

        name = 'ssp_m3_submissions'
        settings = {
            'number_of_shards': 1,
            'number_of_replicas': 0,
        }

    class Django:
        """Django model reference and field mapping."""

        model = SSP_M3
        fields = [
            'RecordType',
            'RPT_MONTH_YEAR',
            'CASE_NUMBER',
            'FIPS_CODE',
            'FAMILY_AFFILIATION',
            'DATE_OF_BIRTH',
            'SSN',
            'RACE_HISPANIC',
            'RACE_AMER_INDIAN',
            'RACE_ASIAN',
            'RACE_BLACK',
            'RACE_HAWAIIAN',
            'RACE_WHITE',
            'GENDER',
            'RECEIVE_NONSSI_BENEFITS',
            'RECEIVE_SSI',
            'RELATIONSHIP_HOH',
            'PARENT_MINOR_CHILD',
            'EDUCATION_LEVEL',
            'CITIZENSHIP_STATUS',
            'UNEARNED_SSI',
            'OTHER_UNEARNED_INCOME',
        ]

@registry.register_document
class SSP_M6DataSubmissionDocument(DocumentBase, Document):
    """Elastic search model mapping for a parsed SSP M6 data file."""

    class Index:
        """ElasticSearch index generation settings."""

        name = 'ssp_m6_submissions'
        settings = {
            'number_of_shards': 1,
            'number_of_replicas': 0,
        }

    class Django:
        """Django model reference and field mapping."""

        model = SSP_M6
        fields = [
            'RecordType',
            'CALENDAR_QUARTER',
            'RPT_MONTH_YEAR',
            'SSPMOE_FAMILIES',
            'NUM_2_PARENTS',
            'NUM_1_PARENTS',
            'NUM_NO_PARENTS',
            'NUM_RECIPIENTS',
            'ADULT_RECIPIENTS',
            'CHILD_RECIPIENTS',
            'NONCUSTODIALS',
            'AMT_ASSISTANCE',
            'CLOSED_CASES',
        ]

@registry.register_document
class SSP_M7DataSubmissionDocument(DocumentBase, Document):
    """Elastic search model mapping for a parsed SSP M7 data file."""

    class Index:
        """ElasticSearch index generation settings."""

        name = 'ssp_m7_submissions'
        settings = {
            'number_of_shards': 1,
            'number_of_replicas': 0,
        }

    class Django:
        """Django model reference and field mapping."""

        model = SSP_M7
        fields = [
            "RecordType",
            "CALENDAR_QUARTER",
            "RPT_MONTH_YEAR",
            "TDRS_SECTION_IND",
            "STRATUM",
            "FAMILIES_MONTH",
        ]

File: tdrs-backend/tdpservice/parsers/go-parser/internal/writer/convert/ssp.go
package convert

import (
	"go-parser/internal/parser"
)

// SSP M1 Converter - Case-level data for active SSP cases

func convertSspM1(record *parser.ParsedRecord, datafileID int32) [][]any {
	return singleRow([]any{
		toText(record.Get("RecordType")),
		toInt4(record.Get("RPT_MONTH_YEAR")),
		toText(record.Get("CASE_NUMBER")),
		toText(record.Get("FIPS_CODE")),
		toText(record.Get("COUNTY_FIPS_CODE")),
		toText(record.Get("STRATUM")),
		toText(record.Get("ZIP_CODE")),
		toInt4(record.Get("DISPOSITION")),
		toInt4(record.Get("NBR_FAMILY_MEMBERS")),
		toInt4(record.Get("FAMILY_TYPE")),
		toInt4(record.Get("TANF_ASST_IN_6MONTHS")),
		toInt4(record.Get("RECEIVES_SUB_HOUSING")),
		toInt4(record.Get("RECEIVES_MED_ASSISTANCE")),
		toInt4(record.Get("RECEIVES_FOOD_STAMPS")),
		toInt4(record.Get("AMT_FOOD_STAMP_ASSISTANCE")),
		toInt4(record.Get("RECEIVES_SUB_CC")),
		toInt4(record.Get("AMT_SUB_CC")),
		toInt4(record.Get("CHILD_SUPPORT_AMT")),
		toInt4(record.Get("FAMILY_CASH_RESOURCES")),
		toInt4(record.Get("CASH_AMOUNT")),
		toInt4(record.Get("NBR_MONTHS")),
		toInt4(record.Get("CC_AMOUNT")),
		toInt4(record.Get("CHILDREN_COVERED")),
		toInt4(record.Get("CC_NBR_MONTHS")),
		toInt4(record.Get("TRANSP_AMOUNT")),
		toInt4(record.Get("TRANSP_NBR_MONTHS")),
		toInt4(record.Get("TRANSITION_SERVICES_AMOUNT")),
		toInt4(record.Get("TRANSITION_NBR_MONTHS")),
		toInt4(record.Get("OTHER_AMOUNT")),
		toInt4(record.Get("OTHER_NBR_MONTHS")),
		toInt4(record.Get("SANC_REDUCTION_AMT")),
		toInt4(record.Get("WORK_REQ_SANCTION")),
		toInt4(record.Get("FAMILY_SANC_ADULT")),
		toInt4(record.Get("SANC_TEEN_PARENT")),
		toInt4(record.Get("NON_COOPERATION_CSE")),
		toInt4(record.Get("FAILURE_TO_COMPLY")),
		toInt4(record.Get("OTHER_SANCTION")),
		toInt4(record.Get("RECOUPMENT_PRIOR_OVRPMT")),
		toInt4(record.Get("OTHER_TOTAL_REDUCTIONS")),
		toInt4(record.Get("FAMILY_CAP")),
		toInt4(record.Get("REDUCTIONS_ON_RECEIPTS")),
		toInt4(record.Get("OTHER_NON_SANCTION")),
		toInt4(record.Get("WAIVER_EVAL_CONTROL_GRPS")),
		newUUID(),
		toDatafileID(datafileID),
		toLineNumber(record.LineNumber),
	})
}

// SSP M2 Converter - Adult-level data for active SSP cases

func convertSspM2(record *parser.ParsedRecord, datafileID int32) [][]any {
	return singleRow([]any{
		toText(record.Get("RecordType")),
		toInt4(record.Get("RPT_MONTH_YEAR")),
		toText(record.Get("CASE_NUMBER")),
		toText(record.Get("FIPS_CODE")),
		toInt4(record.Get("FAMILY_AFFILIATION")),
		toInt4(record.Get("NONCUSTODIAL_PARENT")),
		toText(record.Get("DATE_OF_BIRTH")),
		toText(record.Get("SSN")),
		toInt4(record.Get("RACE_HISPANIC")),
		toInt4(record.Get("RACE_AMER_INDIAN")),
		toInt4(record.Get("RACE_ASIAN")),
		toInt4(record.Get("RACE_BLACK")),
		toInt4(record.Get("RACE_HAWAIIAN")),
		toInt4(record.Get("RACE_WHITE")),
		toInt4(record.Get("SEX")),
		toInt4(record.Get("FED_OASDI_PROGRAM")),
		toInt4(record.Get("FED_DISABILITY_STATUS")),
		toInt4(record.Get("DISABLED_TITLE_XIVAPDT")),
		toInt4(record.Get("AID_AGED_BLIND")),
		toInt4(record.Get("RECEIVE_SSI")),
		toInt4(record.Get("MARITAL_STATUS")),
		toInt4(record.Get("RELATIONSHIP_HOH")),
		toInt4(record.Get("PARENT_MINOR_CHILD")),
		toInt4(record.Get("NEEDS_PREGNANT_WOMAN")),
		toText(record.Get("EDUCATION_LEVEL")),
		toInt4(record.Get("CITIZENSHIP_STATUS")),
		toInt4(record.Get("COOPERATION_CHILD_SUPPORT")),
		toInt4(record.Get("EMPLOYMENT_STATUS")),
		toInt4(record.Get("WORK_ELIGIBLE_INDICATOR")),
		toInt4(record.Get("WORK_PART_STATUS")),
		toInt4(record.Get("UNSUB_EMPLOYMENT")),
		toInt4(record.Get("SUB_PRIVATE_EMPLOYMENT")),
		toInt4(record.Get("SUB_PUBLIC_EMPLOYMENT")),
		toInt4(record.Get("WORK_EXPERIENCE_HOP")),
		toInt4(record.Get("WORK_EXPERIENCE_EA")),
		toInt4(record.Get("WORK_EXPERIENCE_HOL")),
		toInt4(record.Get("OJT")),
		toInt4(record.Get("JOB_SEARCH_HOP")),
		toInt4(record.Get("JOB_SEARCH_EA")),
		toInt4(record.Get("JOB_SEARCH_HOL")),
		toInt4(record.Get("COMM_SERVICES_HOP")),
		toInt4(record.Get("COMM_SERVICES_EA")),
		toInt4(record.Get("COMM_SERVICES_HOL")),
		toInt4(record.Get("VOCATIONAL_ED_TRAINING_HOP")),
		toInt4(record.Get("VOCATIONAL_ED_TRAINING_EA")),
		toInt4(record.Get("VOCATIONAL_ED_TRAINING_HOL")),
		toInt4(record.Get("JOB_SKILLS_TRAINING_HOP")),
		toInt4(record.Get("JOB_SKILLS_TRAINING_EA")),
		toInt4(record.Get("JOB_SKILLS_TRAINING_HOL")),
		toInt4(record.Get("ED_NO_HIGH_SCHOOL_DIPL_HOP")),
		toInt4(record.Get("ED_NO_HIGH_SCHOOL_DIPL_EA")),
		toInt4(record.Get("ED_NO_HIGH_SCHOOL_DIPL_HOL")),
		toInt4(record.Get("SCHOOL_ATTENDENCE_HOP")),
		toInt4(record.Get("SCHOOL_ATTENDENCE_EA")),
		toInt4(record.Get("SCHOOL_ATTENDENCE_HOL")),
		toInt4(record.Get("PROVIDE_CC_HOP")),
		toInt4(record.Get("PROVIDE_CC_EA")),
		toInt4(record.Get("PROVIDE_CC_HOL")),
		toInt4(record.Get("OTHER_WORK_ACTIVITIES")),
		toInt4(record.Get("DEEMED_HOURS_FOR_OVERALL")),
		toInt4(record.Get("DEEMED_HOURS_FOR_TWO_PARENT")),
		toInt4(record.Get("EARNED_INCOME")),
		toInt4(record.Get("UNEARNED_INCOME_TAX_CREDIT")),
		toInt4(record.Get("UNEARNED_SOCIAL_SECURITY")),
		toInt4(record.Get("UNEARNED_SSI")),
		toInt4(record.Get("UNEARNED_WORKERS_COMP")),
		toInt4(record.Get("OTHER_UNEARNED_INCOME")),
		newUUID(),
		toDatafileID(datafileID),
		toLineNumber(record.LineNumber),
	})
}

// SSP M3 Converter - Child-level data for active SSP cases

func convertSspM3(record *parser.ParsedRecord, datafileID int32) [][]any {
	return singleRow([]any{
		toText(record.Get("RecordType")),
		toInt4(record.Get("RPT_MONTH_YEAR")),
		toText(record.Get("CASE_NUMBER")),
		toText(record.Get("FIPS_CODE")),
		toInt4(record.Get("FAMILY_AFFILIATION")),
		toText(record.Get("DATE_OF_BIRTH")),
		toText(record.Get("SSN")),
		toInt4(record.Get("RACE_HISPANIC")),
		toInt4(record.Get("RACE_AMER_INDIAN")),
		toInt4(record.Get("RACE_ASIAN")),
		toInt4(record.Get("RACE_BLACK")),
		toInt4(record.Get("RACE_HAWAIIAN")),
		toInt4(record.Get("RACE_WHITE")),
		toInt4(record.Get("SEX")),
		toInt4(record.Get("RECEIVE_NONSSI_BENEFITS")),
		toInt4(record.Get("RECEIVE_SSI")),
		toInt4(record.Get("RELATIONSHIP_HOH")),
		toInt4(record.Get("PARENT_MINOR_CHILD")),
		toText(record.Get("EDUCATION_LEVEL")),
		toInt4(record.Get("CITIZENSHIP_STATUS")),
		toInt4(record.Get("UNEARNED_SSI")),
		toInt4(record.Get("OTHER_UNEARNED_INCOME")),
		newUUID(),
		toDatafileID(datafileID),
		toLineNumber(record.LineNumber),
	})
}

// SSP M4 Converter - Case-level data for closed SSP cases

func convertSspM4(record *parser.ParsedRecord, datafileID int32) [][]any {
	return singleRow([]any{
		toText(record.Get("RecordType")),
		toInt4(record.Get("RPT_MONTH_YEAR")),
		toText(record.Get("CASE_NUMBER")),
		toText(record.Get("COUNTY_FIPS_CODE")),
		toText(record.Get("STRATUM")),
		toText(record.Get("ZIP_CODE")),
		toInt4(record.Get("DISPOSITION")),
		toText(record.Get("CLOSURE_REASON")),
		toInt4(record.Get("REC_SUB_HOUSING")),
		toInt4(record.Get("REC_MED_ASSIST")),
		toInt4(record.Get("REC_FOOD_STAMPS")),
		toInt4(record.Get("REC_SUB_CC")),
		newUUID(),
		toDatafileID(datafileID),
		toLineNumber(record.LineNumber),
	})
}

// SSP M5 Converter - Adult-level data for closed SSP cases

func convertSspM5(record *parser.ParsedRecord, datafileID int32) [][]any {
	return singleRow([]any{
		toText(record.Get("RecordType")),
		toInt4(record.Get("RPT_MONTH_YEAR")),
		toText(record.Get("CASE_NUMBER")),
		toInt4(record.Get("FAMILY_AFFILIATION")),
		toText(record.Get("DATE_OF_BIRTH")),
		toText(record.Get("SSN")),
		toInt4(record.Get("RACE_HISPANIC")),
		toInt4(record.Get("RACE_AMER_INDIAN")),
		toInt4(record.Get("RACE_ASIAN")),
		toInt4(record.Get("RACE_BLACK")),
		toInt4(record.Get("RACE_HAWAIIAN")),
		toInt4(record.Get("RACE_WHITE")),
		toInt4(record.Get("SEX")),
		toInt4(record.Get("REC_OASDI_INSURANCE")),
		toInt4(record.Get("REC_FEDERAL_DISABILITY")),
		toInt4(record.Get("REC_AID_TOTALLY_DISABLED")),
		toInt4(record.Get("REC_AID_AGED_BLIND")),
		toInt4(record.Get("REC_SSI")),
		toInt4(record.Get("MARITAL_STATUS")),
		toText(record.Get("RELATIONSHIP_HOH")),
		toInt4(record.Get("PARENT_MINOR_CHILD")),
		toInt4(record.Get("NEEDS_OF_PREGNANT_WOMAN")),
		toText(record.Get("EDUCATION_LEVEL")),
		toInt4(record.Get("CITIZENSHIP_STATUS")),
		toInt4(record.Get("EMPLOYMENT_STATUS")),
		toText(record.Get("AMOUNT_EARNED_INCOME")),
		toText(record.Get("AMOUNT_UNEARNED_INCOME")),
		newUUID(),
		toDatafileID(datafileID),
		toLineNumber(record.LineNumber),
	})
}

// SSP M6 Converter - Aggregate data

func convertSspM6(record *parser.ParsedRecord, datafileID int32) [][]any {
	return singleRow([]any{
		toText(record.Get("RecordType")),
		toInt4(record.Get("CALENDAR_QUARTER")),
		toInt4(record.Get("RPT_MONTH_YEAR")),
		toInt4(record.Get("SSP_MOE_FAMILIES")),
		toInt4(record.Get("NUM_2_PARENTS")),
		toInt4(record.Get("NUM_1_PARENTS")),
		toInt4(record.Get("NUM_NO_PARENTS")),
		toInt4(record.Get("NUM_RECIPIENTS")),
		toInt4(record.Get("ADULT_RECIPIENTS")),
		toInt4(record.Get("CHILD_RECIPIENTS")),
		toInt4(record.Get("NONCUSTODIALS")),
		toInt4(record.Get("AMT_ASSISTANCE")),
		toInt4(record.Get("CLOSED_CASES")),
		newUUID(),
		toDatafileID(datafileID),
		toLineNumber(record.LineNumber),
	})
}

// SSP M7 Converter - Stratum data

func convertSspM7(record *parser.ParsedRecord, datafileID int32) [][]any {
	return singleRow([]any{
		toText(record.Get("RecordType")),
		toInt4(record.Get("CALENDAR_QUARTER")),
		toInt4(record.Get("RPT_MONTH_YEAR")),
		toText(record.Get("TDRS_SECTION_IND")),
		toText(record.Get("STRATUM")),
		toInt4(record.Get("FAMILIES_MONTH")),
		newUUID(),
		toDatafileID(datafileID),
		toLineNumber(record.LineNumber),
	})
}
